tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,173 @@
+//! Windows named pipes.
+
+use mio::windows::NamedPipe as MioNamedPipe;
+use winapi::um::winbase::FILE_FLAG_OVERLAPPED;
+
+use std::{
+    ffi::OsStr,
+    fs::OpenOptions,
+    io::{ErrorKind, IoSlice, Result},
+    os::windows::prelude::*,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use crate::io::{AsyncRead, AsyncWrite, Interest, PollEvented, ReadBuf, Ready};
+
+/// Non-blocking windows named pipe.
+#[derive(Debug)]
+pub struct NamedPipe {
+    io: PollEvented<MioNamedPipe>,
+}
+
+impl NamedPipe {
+    fn from_mio(pipe: MioNamedPipe) -> Result<Self> {
+        let io = PollEvented::new(pipe)?;
+        Ok(Self { io })
+    }
+
+    /// Binds new pipe server to the given pipe name.
+    ///
+    /// # Errors
+    ///
+    /// I'll error if there exists another instance of a server bound to that name.
+    fn bind<A: AsRef<OsStr>>(addr: A) -> Result<Self> {
+        let pipe = MioNamedPipe::new(addr)?;
+        Self::from_mio(pipe)
+    }
+
+    /// Opens given pipe as a client.
+    ///
+    /// # Errors
+    ///
+    /// I'll error if there is no such pipe or if there is no servers available.
+    fn open<A: AsRef<OsStr>>(addr: A) -> Result<Self> {
+        let file = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .custom_flags(FILE_FLAG_OVERLAPPED)
+            .open(addr.as_ref())?;
```

Comment:
If I understand correctly, the goal here is to omit the `FILE_FLAG_FIRST_PIPE_INSTANCE` to make this open-only and not create-only?

We should probably add this to mio at some point.

Reply:
> We should probably add this to mio at some point.

Yeah, I meant exactly this saying that mio's interface limits the miow named pipes implementation (miow [supports this flag][1]). I wonder if it's OK to simply add `From<miow::NamedPIpe> for NamedPipe` into mio to not to rewrite miow's `NamedPipeBuilder` 🤔. 

---

Regarding this `open` call – I believe `FILE_FLAG_FIRST_PIPE_INSTANCE` is irrelevant here, since this call leads to `OpenFile2`.

[1]: https://github.com/yoshuawuyts/miow/blob/master/src/pipe.rs#L493-L496

Reply:
> Regarding this open call – I believe FILE_FLAG_FIRST_PIPE_INSTANCE is irrelevant here, since this call leads to OpenFile2.

Apologies for the confusion. I was asking to clarify if I read correctly that the intent with creating the file object ourselves was to avoid this flag. [Since miow would otherwise set it](https://docs.rs/miow/0.3.6/src/miow/pipe.rs.html#468) (and we don't have a way of controlling that through mio).